const express = require("express");
const cors =  require('cors');

const connectToDB = require("./database/db");

process.on("uncaughtException", (error) => {
    console.log("Uncaught Exception..... ðŸ’£ ðŸ”¥ stopping the server....");
    console.log(error.name, error.message);

    process.exit(1);
});

const app = express();

const ErrorsMiddleware = require("./middleware/errorMiddleware");
const LibraryError = require("./utils/libraryError");
const bookRoutes = require("./routes/bookRoutes");

app.use(cors());
app.use(express.json());
app.use(express.urlencoded({ extended: true }));
app.use(ErrorsMiddleware);
app.use("/api/v1/", bookRoutes);

connectToDB();

app.get("/", (req, res) => {
    res.send("books App");
});
app.all("*", (req, res, next) => {
    next(
        new LibraryError(`Can't find ${req.originalUrl} on this server!`, 404)
    );
});
app.listen(8080, () => {
    console.log("listening on 8080");
});

process.on("unhandledRejection", (error) => {
    console.log("Unhandled Rejection..... ðŸ’£ ðŸ”¥ stopping the server....");
    console.log(error.name, error.message);
    server.close(() => {
        // exit code 1 means that there is an issue that caused the program to exit
        process.exit(1);
    });
});
